walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,103 @@
 
 package com.google.sps;
 
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.List;
+import java.util.ListIterator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Cases
+    // Options for no attendees - should be available all day.
+    if (request.getAttendees().isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    // Duration longer than a day will return empty array.
+    if (request.getDuration() > TimeRange.getTimeInMinutes(23, 59)) {
+      return Arrays.asList();
+    }
+    // Check if there are availabilities for all attendees
+    List<String> attendees = new ArrayList<>();
+    attendees.addAll(request.getAttendees());
+    attendees.addAll(request.getOptionalAttendees());
+    Collection<TimeRange> allAttendeesAvailabilities = findAvailability(events, request, attendees);
+
+    // If there are no available timeslots for all attendees including optional attendees,
+    // Find availabilities for the mandatory attendees.
+    if (allAttendeesAvailabilities.isEmpty()) {
+      return findAvailability(events, request, request.getAttendees());
+    }
+
+    return allAttendeesAvailabilities;
+  }
+
+  /** Private helper method to find all availabilities of people from the request. */
+  private Collection<TimeRange> findAvailability(
+      Collection<Event> events, MeetingRequest request, Collection<String> attendees) {
+    // Find all the unavailable times of the attendees.
+    List<TimeRange> unavailabilities = new ArrayList<>();
+    for (Event event : events) {
+      if (!Collections.disjoint(attendees, event.getAttendees())) {
+        unavailabilities.add(event.getWhen());
+      }
+    }
```

Comment:
FYI: This can be written more elegantly with streams.

Reply:
as in `if (attendees.stream().anyMatch(s -> event.getAttendees().contains(s))) {`?

Reply:
no.

unavailabilities = event.stream().filter(...).map(...).collect(...).